### ISSUE TITLE
[infra] Android SDK+NDK CIPD packages should follow the standard directory structure.

### ISSUE THREAD

**Issue by dcharkes** [flutter/flutter#136666]
TL;DR: The CI bots have the Android `sdk` and `ndk` folders as siblings, and set `ANDROID_SDK_ROOT` and `ANDROID_NDK_PATH`. An env-var for the NDK is non-standard, and the standard location of the `ndk` directory is nested inside the `sdk` directory. We should update our DEPS and scripts to do the same.

Long version:

The Android SDK & NDK are installed through a CIPD package: https://chrome-infra-packages.appspot.com/p/flutter/android/sdk/all
The directory structure inside the package (for every OS) is the `ndk` and `sdk` directories as siblings.

Our infra sets the `ANDROID_NDK_PATH` env var https://cs.opensource.google/flutter/recipes/+/main:recipe_modules/flutter_deps/api.py;l=341, and presumably we use this env var somewhere in our build (but I have not been able to find uses in flutter repos including the recipes).

When the Android SDK & NDK are installed on a users' machine through Android Studio or the SDK manager, the `sdk` directory contains the `ndk` directory. (Older SDK versions had an `ndk-bundle` directory.)

The official Android documentation makes no mention of using an environment variable for the NDK location: https://developer.android.com/studio/command-line/variables.html#envar

An online search reveals that project use conflicting env vars for the NDK: `ANDROID_NDK_PATH`, `ANDROID_NDK_HOME`, and `ANDROID_NDK_ROOT`. The latter seems the most common. But none of them is mentioned in the official Android documentation on recommendations.

https://github.com/flutter/flutter/pull/135148 adds logic for NDK discovery on top of the SDK discovery. Since there is no well defined env variable and the ndk is always nested as `ndk/` inside `sdk/` (old versions of the Android SDK can contain `ndk-bundle` instead of `ndk`), and we'd like to use an NDK belonging to the SDK we're using, we look for a nested `ndk` directory inside `sdk`.

This fails on our CI bots currently, because that `ndk/` directory is not there. Our script for making the CIPD package explicitly moves the `ndk` directory out of the `sdk` directory: 

```sh
  # Special treatment for NDK to move to expected directory.
  mv $upload_dir/sdk/ndk $upload_dir/ndk-bundle
  ndk_sub_paths=`find $upload_dir/ndk-bundle -maxdepth 1 -type d`
  ndk_sub_paths_arr=($ndk_sub_paths)
  mv ${ndk_sub_paths_arr[1]} $upload_dir/ndk
  rm -rf $upload_dir/ndk-bundle
```

https://github.com/flutter/engine/blob/c24f30343cf6506bda6a56b833e77461f299fd0c/tools/android_sdk/create_cipd_packages.sh#L118-L123

Changing the CIPD package to _not_ modify the directory structure makes the tests on the above mentioned PR pass ([log](https://logs.chromium.org/logs/flutter/led/godofredoc_google.com/0ee7e718653360504020ce3cc89cd1bd1a312acc8c5ab8a2fb87b008eca8bac5/+/u/run_test.dart_for_tool_integration_tests_shard_and_subshard_5_6/test_stdout)). ([log of failing tests](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8767530688386819265/+/u/run_test.dart_for_tool_integration_tests_shard_and_subshard_5_6/test_stdout))

Changing the directory structure will likely require multiple changes:
* The recipes to set up `ANDROID_NDK_PATH` correctly again ([link](https://cs.opensource.google/flutter/recipes/+/main:recipe_modules/flutter_deps/api.py;l=341)).
* The script to make the CIPD package ([link](https://github.com/flutter/engine/blob/c24f30343cf6506bda6a56b833e77461f299fd0c/tools/android_sdk/create_cipd_packages.sh#L118-L123)).
* Stuff that pops up when trying to make this change.

Many thanks to @godofredoc for helping me triage this! 🙇 

@reidbaker Do you agree we should depend the `ndk` directory being inside the `sdk` directory in the discovery in `lib/src/android/android_sdk.dart` in https://github.com/flutter/flutter/pull/135148?

@GaryQian You have modified the mentioned lines in create_cipd_packages.sh. Would you happen to know why we have a non-standard directory structure? And any concerns about the above proposal?

**Comment by reidbaker** [comment]
Gary no longer works on flutter. @gmackall however might be who you intended to tag.

**Comment by godofredoc** [comment]
@zanderso do you know what would be the complexity of updating the engine build rules to use ndk from a nested folder in sdk?

**Comment by zanderso** [comment]
The script that creates the package would need to be updated here: https://github.com/flutter/engine/blob/main/tools/android_sdk/create_cipd_packages.sh

After that, it's probably just a matter of updating paths like here https://github.com/flutter/buildroot/blob/master/build/config/android/config.gni#L81

and here

https://github.com/flutter/engine/blob/main/build/secondary/third_party/android_tools/BUILD.gn

**Comment by dcharkes** [comment]
@godofredoc Any progress? Or should I take a stab at this myself?

**Comment by dcharkes** [comment]
Update, `sdk/ndk/` doesn't cover it correctly, it should be `sdk/ndk/<version-number>/`.

> The script that creates the package would need to be updated here: https://github.com/flutter/engine/blob/main/tools/android_sdk/create_cipd_packages.sh

I'll take a stab at updating the script tomorrow.

**Comment by dcharkes** [comment]
> The script that creates the package would need to be updated here: https://github.com/flutter/engine/blob/main/tools/android_sdk/create_cipd_packages.sh

https://github.com/flutter/engine/pull/47454

> After that, it's probably just a matter of updating paths like here https://github.com/flutter/buildroot/blob/master/build/config/android/config.gni#L81
> 
> and here
> 
> https://github.com/flutter/engine/blob/main/build/secondary/third_party/android_tools/BUILD.gn

We'll need some conditional logic in there to find the version folder inside the `sdk/ndk/xx.xxx/` (to avoid hardcoding the version number there and having to update the DEPS in tandem with that version number).

(@godofredoc pointed to a file in the recipes repo earlier, but that logic should have all been merged into the GN files.)

**Comment by dcharkes** [comment]
An investigation of our build files and helper scripts shows that we have quite many places where we have the paths hardcoded:

* https://github.com/flutter/engine/blob/a47c721db226c4b57082ca5c389c73843f996f70/build/secondary/third_party/android_tools/BUILD.gn#L8-L14
* https://github.com/flutter/engine/blob/a47c721db226c4b57082ca5c389c73843f996f70/ci/binary_size_treemap.sh#L25
* https://github.com/flutter/engine/blob/a47c721db226c4b57082ca5c389c73843f996f70/sky/tools/flutter_gdb#L89
* https://github.com/flutter/engine/blob/a47c721db226c4b57082ca5c389c73843f996f70/testing/scenario_app/run_android_tests.sh#L49
* https://github.com/flutter/engine/blob/a47c721db226c4b57082ca5c389c73843f996f70/testing/analyze_core_dump.sh#L18

Moreover, the Flutter GN files import the Dart GN files:

* https://github.com/flutter/engine/blob/a47c721db226c4b57082ca5c389c73843f996f70/build/secondary/third_party/android_tools/BUILD.gn#L5
* https://github.com/dart-lang/sdk/blob/75c975fe6d964f08cf03fe13c63a5238f4a28865/build/config/android/rules.gni#L8
* https://github.com/dart-lang/sdk/blob/75c975fe6d964f08cf03fe13c63a5238f4a28865/build/config/android/config.gni#L57

Which also hardcodes the `ndk` folder.

Flutter and Dart don't agree on the NDK version to build and test with. o, we can't change the path from `ndk` to `sdk/ndk/22.x.xxxxxxx/`. In Dart we are already on NDK `27.x.xxxxxxx`, because we're testing Windows arm64 and RISC-V.

This makes me think [the CIPD package was shuffling the directory structure](https://github.com/flutter/engine/pull/47454) for a good reason: To avoid the hassle in all places dealing with a dynamic directory that changes every version change and allow for running the build with different versions (e.g. the flutter build reuses Dart GN rules but uses a different NDK version).

So, maybe we should go back and reconsider `flutter_tools` checking if `ANDROID_NDK_ROOT`, `ANDROID_NDK_PATH`, or `ANDROID_NDK_HOME`  is provided, and only if these are not provided, try to find an NDK inside `$sdkPath/ndk/xx.x.xxxxxxx`.

**Comment by zanderso** [comment]
> Moreover, the Flutter GN files import the Dart GN files:

For the engine GN files that refer to the NDK and set up the Android build, I don't think they share any logic or import files from the Dart build. The build roots are separate. Did you actually run into any issues related to this?

Also, @reidbaker and will soon be conspiring to update the Engine's NDK version.

**Comment by dcharkes** [comment]
> For the engine GN files that refer to the NDK and set up the Android build, I don't think they share any logic or import files from the Dart build. The build roots are separate. Did you actually run into any issues related to this?

Maybe I was misreading the GN files https://github.com/flutter/engine/blob/a47c721db226c4b57082ca5c389c73843f996f70/build/secondary/third_party/android_tools/BUILD.gn#L5:

```
import("//build/config/android/rules.gni")
```

This file doesn't exist in the flutter engine repo. I now see it exists both in the dart and buildroot repo. I hadn't noticed the latter earlier.

Okay, so we could modify the GN files between the Dart and Flutter repo independently.

That still leaves the question open if we would want to modify the the GN files on roles.

> Also, @reidbaker and will soon be conspiring to update the Engine's NDK version.

That has already happened, it's at 34v3 instead of 33v6 now.

**Comment by zanderso** [comment]
> > Also, @reidbaker and will soon be conspiring to update the Engine's NDK version.
> 
> That has already happened, it's at 34v3 instead of 33v6 now.

I think we might be referring to different things. Any more recent version of the NDK than the one we have now has some large changes to layout, and removes support for building with gcc, so I'm pretty sure that hasn't happened yet.

**Comment by dcharkes** [comment]
From my POV, https://github.com/flutter/flutter/pull/135148 is unblocked by relying on an env var that points to the NDK instead.

If it would gain us benefit to un-shuffle the NDK directory for other reasons, we can keep this issue open to track that work. If no-one wants to un-shuffle the directory structure, we can close this issue as wont fix.

**Comment by gmackall** [comment]
I'm going to re-open this as it is a valid request independently of its relation to native asset support.

In particular, after https://github.com/flutter/flutter/pull/159756 lands, there will be a couple of cases where CI will need to download the NDK due to this abnormal structure. Fixing the structure would prevent these downloads.

**Comment by jmagman** [comment]
NDK not being found in the standard location and downloading it has become a source of flakes: https://github.com/flutter/flutter/issues/176366
